### DISCUSSION TITLE
Handle enter/leave and valid connection API

### DISCUSSION THREAD

**Discussion by cedeber** [xyflow/xyflow#4123] [Ideas]
Currently when you try to connect an edge to a handle, reactflow is calling `isValidConnection()` and from here I want to, and can, show a status bar message like: "i/o are not compatible" or "max 2 connections possible" for instance.

The UX thing is that it would be nice to remove any of those messages when you're not hovering a handle any longer. I currently didn't find any ways of doing it correctly, like reactflow updates css classes of handles. Although we have the `onMouseLeave` on the `Handle` component, it's not the same behavior as reactflow. Reactflow reacts already when you come close to a handle, before the pointer is hovering it.

So, if not available, it would be nice to have a way to allow that. Something like `onTryingToConnectStart` and `onTryingToConnectStop` or whatever better named functions ;-)

**Comment by Rounak-stha** [comment]
hey @cedeber, did you find a solution to this?
I am running into the same issue.

**Comment by KaivnD** [comment]
I have done something like you said before, actually you don't need `isValidConnection`, just handle onConnect event, here is my code:
```ts
    onConnect: (conn) => {
      // handle on connect event, check the connection is valid or not here
      // you can set status or whatever you want before return undefined
      const nodes = get().nodes;
      let edges = get().edges;
      if (isNewConnCircular(nodes, edges, conn)) {
        return;
      }

      const sourceNode = nodes.find((item) => item.id === conn.source);
      const targetNode = nodes.find((item) => item.id === conn.target);
      if (!sourceNode || !targetNode) return;

      const source = sourceNode.data.outputs.find(
        (item) => item.id === conn.sourceHandle
      );
      const target = targetNode.data.inputs.find(
        (item) => item.id === conn.targetHandle
      );
      if (!source || !target) return;
      if (source.kind !== target.kind) return;

      // shouldEdgeRemovedOnConnect check every edge if edge is not match should be drop in this action
      edges = edges.filter(
        (item) => !shouldEdgeRemovedOnConnect(item, conn, source.kind)
      ); 

      set(
        {
          nodes,
          edges,
          edited: true,
        },
        undefined,
        "on-connect"
      );
    }
```
